Fixed ffi loading issue on Mac OS X (Mojave)

mcdejonge opened the pull request:
See comment in code. On Mac OS X Mojave (ffi-lib "wrap-rtmidi") tries to load wrap-rtmidi.dylib from the directory of the file that requires rtmidi instead of from the directory where rtmidi/main.rkt is stored.

One solution is to symlink wrap-rtmidi.dylib in the directory of your script. This is ugly. The other is to use this hack to force Racket to load the ffi-lib from the same directory as main.rkt. Which is also ugly, but more convenient.

mcdejonge committed:
Fixed ffi loading issue on Mac OS X (Mojave)

See comment in code. On Mac OS X Mojave (ffi-lib "wrap-rtmidi") tries to load wrap-rtmidi.dylib from the directory of the file that requires rtmidi instead of from the directory where rtmidi/main.rkt is stored.

One solution is to symlink wrap-rtmidi.dylib in the directory of your script. This is ugly. The other is to use this hack to force Racket to load the ffi-lib from the same directory as main.rkt. Which is also ugly, but more convenient.

jbclements commented:
Is what you've written here different from a use of "define-runtime-path"? If not, I think that would be a lot cleaner.

The "right way" to avoid this whole mess (that is, requiring the user to compile C++ files) would be to build packages for each supported platform, and do a platform-dependent info.rkt dependency, as e.g. in Jens Soegaard's "racket-poppler" package:

https://github.com/soegaard/racket-poppler/blob/master/info.rkt

The hardest one for me to compile for is Windows; what platform are you using?

jbclements commented:
Actually, you wrote ".dylib", so I'm guessing ... it's a mac?

mcdejonge commented:
Hi @jbclements ,

Thanks for replying to a pull request on such an old (but very useful :-) ) package.

To answer your questions:

- define-runtime-path is new to me, being a Racket noob. I'll look into it and send a new pull request if that solves the problem.
- I'm on a Mac, yes, but I've heard the same problem occurs on Linux.
- I'm not on Windows, so I can't help you there, sorry.

If building packages for supported platforms is feasible, like in the racket-poppler package, that would be a better solution. I'll take a look at it, though that might take a while. 

Anyway, like I said I'll look into making a cleaner fix with define-runtime-path and send you a new pull request, safe in the knowledge that this package still has enough life in it that it will be incorporated at some point.

For now, thank you for spending time on this.

mcdejonge committed:
Load wrap-rtmidi using define-runtime-path

Previous fix was needlessly verbose. @jbclements had the much better idea to use define-runtime-path.

mcdejonge commented:
Ok ... using define-runtime-path is indeed much cleaner. I've updated the pull request. If you could incorporate it in your package that would be great.

Will add working on the much better supported platforms solutions to my list, though.

jbclements commented:
LGTM. Thanks for your patience!
